### PR TITLE
 Do not delete e2e test resources on PR builds

### PR DESCRIPTION
**Pull request by vinagesh**
This PR mainly moves the configuration of which devices to delete to the pipeline variable rather than a constant value. Once this is checked in, the scheduled run will run as normal, running all tests and cleaning up both e2e and sample created devices.

Apart from that pipeline, I"ll add a new pipeline that will only delete the sample created devices along with running the samples. This pipeline will be enabled on PR builds.

After this is complete, I will make merge these changes to preview branch and create a PR pipeline for the preview branch as well.

### CODE REVIEW

File: iot-hub/Samples/service/RegistryManagerSample/Parameters.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ internal class Parameters
             'd',
             "DevicePrefix",
             Required = false,
-            Default = "RegistryManagerSample-",
```
Comment (vinagesh) [inline iot-hub/Samples/service/RegistryManagerSample/Parameters.cs:32]: Changing to match prefix. This was not in the delete list earlier but the sample deletes the device so this is just to guard against leaked resources. Currently the hub does not have any devices with this prefix so we don't need to worry about deleting the ones with old prefix.

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs [hunk 2]
```diff
@@ -65,8 +69,29 @@ private async Task CleanupDevices()
                     excludeKeysInExport: true,
                     storageAuthenticationType: StorageAuthenticationType.KeyBased);
 
-            JobProperties exportAllDevicesJob = await _registryManager.ExportDevicesAsync(exportAllDevicesProperties);
+            JobProperties exportAllDevicesJob = null;
```
Comment (vinagesh) [inline iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:72]: Adding retries on jobs to protect against any currently running job. We could change the total wait time if necessary after observing for a while.

File: vsts/vsts.yaml [hunk 3]
```diff
@@ -53,6 +53,10 @@ jobs:
           IOTHUB_PFX_X509_THUMBPRINT: $(IOTHUB-PFX-X509-THUMBPRINT)
           DPS_IDSCOPE: $(DPS-IDSCOPE)
           PROVISIONING_CONNECTION_STRING: $(PROVISIONING-CONNECTION-STRING)
+          # PATH-TO-DEVICE-PREFIX-FOR-DELETION-FILE is a variable configured on the pipeline. The possible values are: csharp_devices_list.csv, csharp_samples_devices_list.csv.
```
Comment (vinagesh) [inline vsts/vsts.yaml:56]: This variable is no longer a constant and no longer extracted from keyvault.

**Review by timtay-microsoft — APPROVED** [review]

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs [hunk 4]
```diff
@@ -37,7 +41,7 @@ public CleanupDevicesSample(RegistryManager rm, BlobContainerClient sc, List<str
         {
             _registryManager = rm ?? throw new ArgumentNullException(nameof(rm));
             _blobContainerClient = sc ?? throw new ArgumentNullException(nameof(sc));
-
+            Console.WriteLine($"Delete devices with prefixes: {JsonConvert.SerializeObject(deleteDevicesWithPrefix)}");
```
Comment (drwill-ms) [inline iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:44]: Why call `JsonConvert.SerializeObject()` on a list of strings?

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs [hunk 5]
```diff
@@ -65,8 +69,29 @@ private async Task CleanupDevices()
...
 
-            JobProperties exportAllDevicesJob = await _registryManager.ExportDevicesAsync(exportAllDevicesProperties);
+            JobProperties exportAllDevicesJob = null;
+            
+            int tryCount = 0;
+            while (true)
+            {
+                try
+                {
+                    exportAllDevicesJob = await _registryManager.ExportDevicesAsync(exportAllDevicesProperties);
```
Comment (drwill-ms) [inline iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:79]:
```suggestion
                    exportAllDevicesJob = await _registryManager.ExportDevicesAsync(exportAllDevicesProperties).ConfigureAwait(false);

```

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs [hunk 6]
```diff
@@ -65,8 +69,29 @@ private async Task CleanupDevices()
...
+                }
+                // Wait for pending jobs to finish.
+                catch (JobQuotaExceededException) when (++tryCount < MaxIterationWait)
+                {
+                    Console.WriteLine($"JobQuotaExceededException... waiting.");
+                    await Task.Delay(WaitDuration).ConfigureAwait(false);
+                }
+            }
 
+            if(exportAllDevicesJob == null)
```
Comment (drwill-ms) [inline iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:90]:
```suggestion
            if (exportAllDevicesJob == null)

```

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs [hunk 7]
```diff
@@ -126,7 +151,29 @@ private async Task CleanupDevices()
...
-                JobProperties importDevicesToBeDeletedJob = await _registryManager.ImportDevicesAsync(importDevicesToBeDeletedProperties);
+
+                JobProperties importDevicesToBeDeletedJob = null;
+
+                tryCount = 0;
+                while (true)
+                {
+                    try
+                    {
+                        importDevicesToBeDeletedJob = await _registryManager.ImportDevicesAsync(importDevicesToBeDeletedProperties);
```
Comment (drwill-ms) [inline iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:162]:
```suggestion
                        importDevicesToBeDeletedJob = await _registryManager
                            .ImportDevicesAsync(importDevicesToBeDeletedProperties)
                            .ConfigureAwait(false);

```

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs (hunk 4 above)
Comment (vinagesh) [reply iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:44]: I just want to print all the values in the list. I could use string.Join or linq but just chose one way to print. If there is an argument for why one way is better that the other, I can change it.

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs (hunk 5 above)
Comment (vinagesh) [reply iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:79]: From what I know, we don't use configureawait in samples.

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs (hunk 7 above)
Comment (vinagesh) [reply iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:162]: From what I know, we don't use ConfigureAwait in samples.

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs (hunk 4 above)
Comment (drwill-ms) [reply iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:44]: Ah, I see. Yeah, I was a bit confused but now I get it. I think string.join would be more straight forward, but I think this is acceptable as well.

File: iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs (hunk 5 above)
Comment (drwill-ms) [reply iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:79]: You are right. We do have `ConfigureAwait()` elsewhere in this file, though, so we should clean that up then.

Comment (vinagesh) [reply iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs:79]: Will clean up!

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**